### PR TITLE
[wip] dotnet cli restore: read IntermediateOutputPath

### PR DESCRIPTION
**Pull request by 0x53A**
You can redirect where nuget and the cli do their stuff, but paket currently ignores that.

This adds two more parameters: one for where the references file should be written and one for the intermediate dir.

The second one is required so paket knows where to copy the nuget.config.

The first one is not yet required - it could just write it to the intermediate directory.

But for https://github.com/fsprojects/Paket/issues/2606 we will need more than one file (one per configuration/tfm) so I added an explicit parameter now.

### CODE REVIEW

File: integrationtests/Paket.IntegrationTests/RestoreSpecs.fs [hunk 1]
```diff
@@ -42,4 +42,55 @@ let ``#2496 Paket fails on projects that target multiple frameworks``() =
...
     | (true, false) -> printfn "all good!"
+    | _ -> failwithf "dotnet restore exitCode %d > 0? or hasErrorOut %b" exitCode hasErrorOut
+    
+[<Test>]
+let ``#xxx dotnet restore writes paket references file to correct obj dir``() = 
+    let project = "ObjDir"
+    let scenario = "i00xxxx-obj-dir"
+    let tmpPaketFolder = (scenarioTempPath scenario) @@ "../.paket"
+    let targetsFile = FullName(__SOURCE_DIRECTORY__ + "../../../src/Paket/embedded/Paket.Restore.targets")
+    let paketExe = FullName(__SOURCE_DIRECTORY__ + "../../../.paket/paket.exe")
```
Comment (0x53A) [inline integrationtests/Paket.IntegrationTests/RestoreSpecs.fs]: This is wrong (and is also wrong in the test above): it copies the checked-in version of paket, not the previously build one.

Tomorrow ...

File: integrationtests/Paket.IntegrationTests/RestoreSpecs.fs [hunk 2]
```diff
@@ -42,4 +42,59 @@ let ``#2496 Paket fails on projects that target multiple frameworks``() =
...
+[<Test>]
+let ``#2642 dotnet restore writes paket references file to correct obj dir``() = 
+    let project = "ObjDir"
+    let scenario = "i002642-obj-dir"
+    let tmpPaketFolder = (scenarioTempPath scenario) @@ "../.paket"
+    let targetsFile = FullName(__SOURCE_DIRECTORY__ + "../../../src/Paket/embedded/Paket.Restore.targets")
+    let paketExe = FullName(__SOURCE_DIRECTORY__ + "../../../bin/paket.exe")
+
+    setEnvironVar "PaketExePath" paketExe    
+    prepare scenario
```
Comment (matthid) [inline integrationtests/Paket.IntegrationTests/RestoreSpecs.fs]: we now have a new prepareSdk function to simplify these tests :)

File: src/Paket/Paket.fsproj [hunk 3]
```diff
@@ -31,6 +31,14 @@
...
     <StartProgram>paket.exe</StartProgram>
     <StartAction>Project</StartAction>
+    <StartArguments>restore --project "C:\source\Paket\integrationtests\scenarios\i00xxxx-obj-dir\temp\ObjDir\ObjDir.csproj" --target-framework net461 --new-sdk-references-file "C:\source\Paket\integrationtests\scenarios\i00xxxx-obj-dir\temp\ObjDir\MyCustomFancy\xxx.paket.ref" --intermediate-dir "C:\source\Paket\integrationtests\scenarios\i00xxxx-obj-dir\temp\ObjDir\MyCustomFancy"</StartArguments>
+    <StartWorkingDirectory>C:\source\Paket\integrationtests\scenarios\i00xxxx-obj-dir\temp\ObjDir</StartWorkingDirectory>
+    <StartArguments>restore</StartArguments>
+    <StartWorkingDirectory>C:\temp\perf\</StartWorkingDirectory>
+    <StartArguments>update</StartArguments>
+    <StartWorkingDirectory>C:\proj\testing\testpaketfailure\</StartWorkingDirectory>
+    <StartArguments>fix-nuspecs files src/Paket.Core.preview3/obj/Paket.Core.1.0.0.nuspec project-file src/Paket.Core.preview3/Paket.Core.fsproj</StartArguments>
+    <StartWorkingDirectory>C:\proj\Paket</StartWorkingDirectory>
```
Comment (matthid) [inline src/Paket/Paket.fsproj]: I do this myself all the time. I don't know why vs is failing so hard here.

Ideally we would gitignore that section of this file

Comment (0x53A) [reply src/Paket/Paket.fsproj]: yeah ... please remind me again after I remove the wip

File: src/Paket.Core/Installation/RestoreProcess.fs [hunk 4]
```diff
@@ -462,7 +463,8 @@ let Restore(dependenciesFileName,projectFile,force,group,referencesFileNames,ign
                     ProjectFile.FindAllProjects root
                     |> Seq.filter (fun proj -> proj.GetToolsVersion() >= 15.0)
                     |> Seq.iter (fun proj ->
-                        RestoreNewSdkProject dependenciesFile lockFile resolved groups proj
+                        let objDir = objDir |> Option.defaultValue (Path.Combine(Path.GetDirectoryName(proj.FileName),"obj"))
```
Comment (0x53A) [inline src/Paket.Core/Installation/RestoreProcess.fs]: I assume this path is called if I just do "paket restore"? Then we would need to crack the project file and manually parse "$(BaseIntermediateOutputPath)". :\

**Comment by 0x53A** [comment]
@matthid Your performance improvements have made this a little bit harder :\

Before, paket was called each time directly from msbuild, so it was easy to just pass-through any MSBuild variables.

As I understand it now, you now do an "external" restore, which is not triggered by msbuild.

So I have no idea how to pass the MSBuild variables, because I really don't want to crack open and parse the project file manually.

**Comment by matthid** [comment]
@0x53A 

> As I understand it now, you now do an "external" restore, which is not triggered by msbuild.

It is triggered on demand:
 - we will do a full restore when we detect that the last restore is not up to date (paket-files)
 - we will do a partial restore when we detect that the project file has changed

If nothing is outdated we will skip calling paket and will just read and forward the files in the `obj/` directory.

**Comment by 0x53A** [comment]
If I run "paket restore", then run "dotnet restore", which of the two will write the obj\paket.referenced files?

My understanding (which may be totally wrong) is:

* previously: "paket restore" only restored the packages directory, "dotnet restore" called paket once for each TFM.
* now: "paket restore" will generate all files, "dotnet restore" afterwards is a nop and would only do something if the project file was changed.

**Comment by matthid** [comment]
@0x53A Yes both is correct. But if you do NOT call `paket restore` and ONLY `dotnet restore` it will do a `paket restore` as well (but only a single time for ALL projects / TFMs)

**Comment by 0x53A** [comment]
And that is precisely my issue: I need to pass through one variable from inside the project file, which I can't, because the restore is "external".

**Comment by matthid** [comment]
Why? From Paket's world view restore is a global operation. I also pass properties (namely the target framework list for example...)

**Comment by 0x53A** [comment]
How do you read the TFM list? What if it was not set directly in the project file, but in a directories.props?

**Comment by 0x53A** [comment]
Basically this seems like a repetition of https://github.com/fsprojects/Paket/issues/2241, which also failed if the property was not set directly in the project file, but in an imported target/props.

**Comment by matthid** [comment]
@0x53A That is not working sadly, because currently I used the paket internal project framework parser. 
Mid-Long term we need to switch to what @enricosada is doing via https://github.com/enricosada/dotnet-proj-info which works but is slow.

**Comment by matthid** [comment]
> Basically this seems like a repetition of #2241, which also failed if the property was not set directly in the project file, but in an imported target/props.

Correct. Same problem and same code...

**Comment by matthid** [comment]
Alternatively we make paket fast enough to always call it in msbuild. But currently when we do it paket starts to become unusable if you try to use it in a solution with more than 7 projects (at least IMHO).

**Comment by 0x53A** [comment]
> Alternatively we make paket fast enough to always call it in msbuild.

I fear that is the only reliable solution.

Assuming (I haven't profiled it yet) that a non-neglible part is JIT, we should guide users to NGEN paket.exe.

**Comment by matthid** [comment]
> Assuming (I haven't profiled it yet) that a non-neglible part is JIT, we should guide users to NGEN paket.exe.

I don't think that will be enough. Also we should probably find an automatic solution. Always calling NGEN manually after updates is not really a viable option oO

**Comment by 0x53A** [comment]
The best option would be AOT, and then the bootstrapper can detect the system platform and download the specific version.

**Comment by 0x53A** [comment]
I should have finished this earlier and added an integration test, then you would have had to deal with his :\

Sorry, I am just slightly unhappy at the moment.

**Comment by matthid** [comment]
I'm sorry, on my defense I told you somewhere (issue or other PR that I planned to change this).

**Comment by 0x53A** [comment]
Ok.

so my first step would be to use the current parser (even with its known issues) to read ``BaseIntermediateOutputPath``, and then at an unspecified date later maybe try to fix the issues for real without regressing the performance too much.

**Comment by theimowski** [comment]
Has this been abandoned as of now? I'd have some interest in this feature as well

**Comment by theimowski** [comment]
bump up @matthid @0x53A I know it's been a while since you last commented on this, however I'd like to know if anything changed in this regards and it's possible to add support for IntermediateOutputPath? Or are we still blocked by the same issues?
I'm happy to help with this feature if necessary

**Comment by 0x53A** [comment]
The issue back then was that paket did one _external_ restore for all frameworks. It would need to do  restores when called from MsBuild so it has proper access to the MsBuild variables.

I do not know if anything changed in that regard.

**Comment by 0x53A** [comment]
And in the forseable future I don't have much time, so I'll close this PR, sorry.

If you want to work on it, then I guess the strategy would be the same - pass IntermediateOutputPath from Paket.Restore.targets to paket.exe, then route it through.

**Comment by theimowski** [comment]
Thanks @0x53A I'll give it a go

**Pull request closed without merging**